Insert text via API

c-martinez opened the pull request:
Implements feature suggested on issue #2897.

Not quite yet finished, but would like to check this is going in the right direction. Feedback welcome!

Gared commented on src/node/handler/APIHandler.js:
```diff
@@ -56,6 +56,7 @@ var version =
   , "listSessionsOfAuthor"      : ["authorID"]
   , "getText"                   : ["padID", "rev"]
   , "setText"                   : ["padID", "text"]
+  , "insertText"                : ["padID", "text", "lineNum"]
```
Add your new API method only to the latest API version

c-martinez commented:
Ah, yes -- I started by adding it to every version while I was figuring it out. Removed.

JohnMcLear commented:
This needs backend tests plz :)

c-martinez commented:
Better?

Gared commented:
Looks like the pad is being destroyed if you remove the "lineNumber" parameter (I get javascript errors after calling the API without this parameter). Please make this parameter necessary.

Gared commented:
Also please set up a "real author" otherwise you'll get the error message "There is no author for authorId: API-bot"

JohnMcLear commented:
Thanks for test coverage @c-martinez -- Please also add coverage for the bits @Gared has discovered

c-martinez commented:
@Gared, thanks for feedback. lineNum now is a required parameter.

I'm adding authorID also as a required parameter -- but I'm not sure how to test this -- `createAuthor` doesn't seem to have tests either, and I'm not sure how to get a authorID to use in the tests. Any tips?

c-martinez commented:
Hi -- added a "real author" (via authorMapper -- not sure if this is the most logical way), and added tests for it as well.

Along the way, added tests for `createAuthor` and `createAuthorIfNotExistsFor`.

comments?

Gared commented on src/node/db/API.js:
```diff
@@ -298,15 +298,91 @@ exports.setText = function(padID, text, callback)
+    //author does not exist
+    if(authorID == false)
+    {
+      callback(new customError("Author not given or given author does not exits","apierror"));
```
propably there is a misspelling in "exits". Maybe "exist"? :-)

Gared commented on src/node/db/API.js:
```diff
@@ -298,15 +298,91 @@ exports.setText = function(padID, text, callback)
+          for (var i = 0; i < lineNum-1; i++) {
+            insertPoint += oldLines[i].length + 1;
+          }
+        } else {
```
Please use a separate line for the brackets and "else" (code formatting)

Gared commented:
Looks good to me so far. Thx for your work :-)
Will do some tests the next days.

c-martinez commented:
Well spotted @Gared , thanks!

Gared commented on src/node/db/API.js:
```diff
@@ -298,15 +298,93 @@ exports.setText = function(padID, text, callback)
+        }
+        else
+        {
+          insertPoint = oldLen;
```
Where is oldLen set? Seems to me that this variable is not defined at this point

Gared replied:
This seems to be the reason why you get an error if you specify a lineNum that is higher than the actual number of lines on this pad

Gared commented on src/node/db/API.js:
```diff
@@ -298,15 +298,93 @@ exports.setText = function(padID, text, callback)
+    if(ERR(err, callback)) return;
+
+    //author does not exist
+    if(authorID == false)
```
Why do you check the authorID? As far as I can see there is no case that the authorID can be 'false'

c-martinez replied:
Good point

Gared commented on src/node/db/API.js:
```diff
@@ -298,15 +298,93 @@ exports.setText = function(padID, text, callback)
+  }
+
+  //check if lineNum is a number
+  if(!isNaN(parseInt(lineNum)))
```
Please do call "parseInt" only once

Gared commented:
Please also documentate your new API method here:
https://github.com/ether/etherpad-lite/blob/9335ae00d0c889bf96482447990968ca3eef6806/doc/api/http_api.md

c-martinez commented:
Hi @Gared , thanks for comments. (Sorry, it looks like you spent quite a bit of time fixing issues which I should have seen before -- really thanks for that). I'll address them and send a new commit.

Gared commented on src/node/db/API.js:
```diff
@@ -298,15 +298,82 @@ exports.setText = function(padID, text, callback)
+
+  var authorMangager = require("./AuthorManager");
+  // Check author exists
+  authorMangager.createAuthorIfNotExistsFor(authorMapper, false, function(err, authorID)
```
In the callback you get the author object and not the authorID

Gared replied:
Is there any special reason why you set the parameter for "name" to "false"?

Gared commented on src/node/db/API.js:
```diff
@@ -298,15 +298,82 @@ exports.setText = function(padID, text, callback)
+      assem.appendOpWithText('+', newText);
+      assem.endDocument();
+      var typedChanges = Changeset.pack(oldLen, newLen, assem.toString(), newText);
+      pad.appendRevision(typedChanges, authorID);
```
Use here author.authorID

Gared commented:
@c-martinez Can you do these changes, please? I want to release a new version of etherpad with this feature.

c-martinez commented:
Hi @Gared -- yes, sorry, I've been distracted with other things. I'll make these changes later today.

c-martinez commented:
Done!

But I get the impression that I'm still doing something not-quite-right... I'm using `createAuthorIfNotExistsFor` but I don't think I should be creating a new author. Is there a checkIfAuthorExistsForMapper function (or something similar?)

Gared commented:
@c-martinez No, there is currently no such function, but I'm fine to use the author mapper to create an author if necessary. So I would merge your PR now.

c-martinez commented:
Would it be a potentially useful feature?

Gared commented:
@c-martinez I think you're right. If we use the authorMapper to create a new author there is no way to insert text as another author or create an author with a name. So I would prefer to replace the parameter ```authorMapper``` with ```authorID```. There should be a check if this authorID exists, but nothing more. What do you think about?

To answer your question: I don't think such a function would be useful, because the authorMapper is a random string created by an application using the API. In my opinion there is no reason to ask etherpad for an authorMapper.

c-martinez commented:
I guess using an authorID makes more sense then.

Gared commented:
@c-martinez Thanks for your quick changes, but I've noticed that there is a problem after applying a text to a pad:
1. Open a pad
2. Write something in the pad
3. Call insertText-API with the authorID of your current user in the Pad
4. Write something in the pad again
5. You get a disconnect in the pad
```
[WARN] client - Error: applySubmittedChangesToBase: no submitted changes to apply -- { errorId: 'pp8WgPgOyFStijvP4rXA',
  msg: 'Error: applySubmittedChangesToBase: no submitted changes to apply',
  url: 'http://localhost:9001/p/test',
  linenumber: 10465,
  userAgent: 'Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:45.0) Gecko/20100101 Firefox/45.0' }
```

It seems that there is a problem with the author and the change being applied to the pad

c-martinez commented:
@Gared -- I'll look into it.

c-martinez commented:
@Gared -- reproduced the issue you mentioned. It seems like the API call itself is working fine, but the browser session gets disconnected or something.

We could avoid this by not using a normal author to insert text via the API (i.e. do not have an authorID as a parameter). Instead we could create a 'special' author (via `createAuthorIfNotExistsFor` for instance), and always use the same author for API calls. This should solve the issue, right? What do you think?

Gared commented:
@c-martinez That would work, but I would really appreciate if we can add text as any author via the API.
I think there is a problem with the authorID in the change, because the added text via the API isn't colored and has no author attributes. This looks like the main problem to me.

fcassin commented:
The problem seems to be that the client receives an ACCEPT_COMMIT message but there was no submittedChangeset created beforehand. This is what causes the uncaught error.

If you have two users connected on the same pad when the API call is made, the user not "impersonated" by the API sees the changes immediatly, but with no author color. Only the impersonated user loses its session after the API call. Maybe a NEW_CHANGES message is required, instead of a ACCEPT_COMMIT message.

c-martinez commented:
Following @Gared suggestion on #2950, this should add the colour to the text according to the colour set by the user. (Maybe need to check if author already has attributes set for this pad).

JohnMcLear commented:
@c-martinez is this something you still need?

c-martinez commented:
Not really -- I am not using etherpad at the moment.
